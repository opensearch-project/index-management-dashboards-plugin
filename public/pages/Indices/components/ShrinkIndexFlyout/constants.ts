/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DEDAULT_ADVANCED_INDEX_SETTINGS = {
  "index.number_of_shards": "1",
  "index.number_of_replicas": "1",
};

export const INDEX_BLOCKS_WRITE_SETTING = "index.blocks.write";
export const INDEX_BLOCKS_READONLY_SETTING = "index.blocks.read_only";
export const INDEX_ROUTING_ALLOCATION_SETTING = "index.routing.allocation.require";
