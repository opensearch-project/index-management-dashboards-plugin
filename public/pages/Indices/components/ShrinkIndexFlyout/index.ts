/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ShrinkIndexFlyout from "./ShrinkIndexFlyout";

export default ShrinkIndexFlyout;
