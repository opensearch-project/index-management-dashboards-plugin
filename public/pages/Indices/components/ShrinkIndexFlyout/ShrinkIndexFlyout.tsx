/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiCallOut,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiLink,
  EuiSpacer,
  EuiTitle,
} from "@elastic/eui";
import React, { Component } from "react";

import { CatIndex } from "../../../../../server/models/interfaces";
import FlyoutFooter from "../../../VisualCreatePolicy/components/FlyoutFooter";
import CustomFormRow from "../../../../components/CustomFormRow";
import FormGenerator, { IField, IFormGeneratorRef } from "../../../../components/FormGenerator";
import { IndexItem } from "../../../../../models/interfaces";
import { SHRINK_DOCUMENTATION_URL, INDEX_SETTINGS_URL } from "../../../../utils/constants";
import {
  DEDAULT_ADVANCED_INDEX_SETTINGS,
  INDEX_NUMBER_OF_SHARDS_SETTING,
  INDEX_BLOCKS_WRITE_SETTING,
  INDEX_BLOCKS_READONLY_SETTING,
  INDEX_BLOCKS_METADATA_SETTING,
  INDEX_ROUTING_ALLOCATION_SETTING,
  TARGET_INDEX_NAME_REQUIRED_MESSAGE,
} from "./constants";
import _ from "lodash";

interface ShrinkIndexProps {
  sourceIndex: CatIndex;
  onClose: () => void;
  onConfirm: (sourceIndexName: string, targetIndexName: string, indexSettings: {}) => void;
  getIndexSettings: (indexName: string, flat: boolean) => Promise<Object>;
}

interface ShrinkIndexState {
  targetIndexName: string;
  indexSettings: IndexItem["settings"];
  targetIndexNameError: string;
  sourceIndexCannotShrinkErrors: string[];
  sourceIndexNotReadyReasons: string[];
}

export default class ShrinkIndexFlyout extends Component<ShrinkIndexProps, ShrinkIndexState> {
  constructor(props: ShrinkIndexProps) {
    super(props);

    this.state = {
      targetIndexName: "",
      indexSettings: DEDAULT_ADVANCED_INDEX_SETTINGS,
      targetIndexNameError: TARGET_INDEX_NAME_REQUIRED_MESSAGE,
      sourceIndexCannotShrinkErrors: [],
      sourceIndexNotReadyReasons: [],
    };
  }

  async componentDidMount() {
    await this.isSourceIndexReady();
  }

  formRef: IFormGeneratorRef | null = null;

  onClickAction = async () => {
    const { sourceIndex, onConfirm } = this.props;
    const { targetIndexName, targetIndexNameError, indexSettings } = this.state;
    if (!!targetIndexNameError) {
      return;
    }

    const result = await this.formRef?.validatePromise();
    if (result?.errors) {
      return;
    }

    onConfirm(sourceIndex.index, targetIndexName, indexSettings);
  };

  isSourceIndexReady = async () => {
    const { sourceIndex, getIndexSettings } = this.props;
    const sourceIndexCannotShrinkErrors = [];
    let reason = "";

    // Show danger errors and disable the Shrink button if the source index is red or closed, or only has one primary shard.
    if (sourceIndex.health == "red") {
      sourceIndexCannotShrinkErrors.push("The index's health status is [red]!");
    }
    if (sourceIndex.pri == "1") {
      sourceIndexCannotShrinkErrors.push("The index has only one primary shard!");
    }
    if (sourceIndex.status == "close") {
      sourceIndexCannotShrinkErrors.push("The index is closed!");
    }
    if (sourceIndexCannotShrinkErrors.length > 0) {
      this.setState({
        sourceIndexCannotShrinkErrors: sourceIndexCannotShrinkErrors,
      });
      return;
    }

    const sourceIndexNotReadyReasons = [];
    const indexSettings = await getIndexSettings(sourceIndex.index, true);

    // Check whether `index.blocks.read_only` or `index.blocks.metadata` is set firstly,
    // because shrink operation will timeout and then the new shrunkend index's shards cannot be allocated.
    if (!!indexSettings && indexSettings.hasOwnProperty(sourceIndex.index)) {
      if (
        !!indexSettings[sourceIndex.index]["settings"][INDEX_BLOCKS_READONLY_SETTING] ||
        !!indexSettings[sourceIndex.index]["settings"][INDEX_BLOCKS_METADATA_SETTING]
      ) {
        reason =
          "Index setting [" +
          INDEX_BLOCKS_READONLY_SETTING +
          "] or [" +
          INDEX_BLOCKS_METADATA_SETTING +
          "] is [true], this will cause the new shrunken index's shards to be unassigned.";
        sourceIndexNotReadyReasons.push(reason);
      }
    }

    // It's better to do shrink when the source index is green.
    if (sourceIndex.health != "green") {
      reason = "The index's health is not green.";
      sourceIndexNotReadyReasons.push(reason);
    }

    // `index.blocks.write` setting is required.
    if (
      !!indexSettings &&
      indexSettings.hasOwnProperty(sourceIndex.index) &&
      !indexSettings[sourceIndex.index]["settings"][INDEX_BLOCKS_WRITE_SETTING]
    ) {
      reason = "Index setting [index.blocks.write] is not [true].";
      sourceIndexNotReadyReasons.push(reason);
    }

    // This check may not be correct in the following situations:
    // 1. the cluster only has one node, so the source index's shards are allocated to the same node.
    // 2. the primary shards of the source index are just allocated to the same node, not manually.
    // 3. the user set `index.routing.rebalance.enable` to `none` and then manually move each shard's copy to one node.
    // In the above situations, the source index does not have a `index.routing.allocation.require._*` setting which can
    // rellocate one copy of every shard to one node, but it can also execute shrinking successfully if other conditions are met.
    // But in most cases, source index always have many shards distributed on different node,
    // so index.routing.allocation.require._*` setting is required.
    // In above, we just show a warning in the page, it does not affect any button or form.
    if (!!indexSettings && indexSettings.hasOwnProperty(sourceIndex.index)) {
      let shardsAllocatedToOneNode = false;
      const settings = indexSettings[sourceIndex.index]["settings"];
      for (let settingKey in settings) {
        if (settingKey.startsWith(INDEX_ROUTING_ALLOCATION_SETTING)) {
          shardsAllocatedToOneNode = true;
          break;
        }
      }
      if (!shardsAllocatedToOneNode) {
        reason = "One copy of every shard should be allocated to one node.";
        sourceIndexNotReadyReasons.push(reason);
      }
    }

    if (sourceIndexNotReadyReasons.length > 0) {
      this.setState({
        sourceIndexNotReadyReasons: sourceIndexNotReadyReasons,
      });
    }
  };

  onTargetIndexNameChange = (value: string) => {
    let err = "";
    if (!value.trim()) {
      err = TARGET_INDEX_NAME_REQUIRED_MESSAGE;
    }
    this.setState({
      targetIndexName: value,
      targetIndexNameError: err,
    });
  };

  isNumberOfShardsValid = (sourceShardsNum: number, targetShardsNum: number): boolean => {
    if (targetShardsNum <= 0 || sourceShardsNum % targetShardsNum != 0) {
      return false;
    }
    return true;
  };

  render() {
    const { onClose } = this.props;
    const { targetIndexName, targetIndexNameError, indexSettings, sourceIndexCannotShrinkErrors, sourceIndexNotReadyReasons } = this.state;

    const formFields: IField[] = [
      {
        rowProps: {
          label: "Number of shards",
          helpText: "The number of primary shards in the new shrunken index.",
        },
        name: "index.number_of_shards",
        type: "Number",
        options: {
          rules: [
            {
              required: true,
              message: "The number of new primary shards must be a positive factor of the number of primary shards in the source index.",
              validator: (rule, value, callback) => {
                if (!this.isNumberOfShardsValid(this.props.sourceIndex.pri, value)) {
                  return Promise.reject("not valid");
                }
                return Promise.resolve();
              },
            },
          ],
          props: {
            "data-test-subj": "numberOfShardsInput",
            min: 1,
            max: Number(this.props.sourceIndex.pri),
          },
        },
      },
    ];

    return (
      <EuiFlyout ownFocus={true} onClose={() => {}} maxWidth={600} size="m" hideCloseButton>
        <EuiFlyoutHeader hasBorder>
          <EuiTitle size="m">
            <h2 id="flyoutTitle"> Shrink index</h2>
          </EuiTitle>
        </EuiFlyoutHeader>
        <EuiFlyoutBody>
          <EuiCallOut color="danger" hidden={sourceIndexCannotShrinkErrors.length == 0}>
            <div style={{ lineHeight: 1.5 }}>
              <p>The source index cannot shrink, due to the following reasons:</p>
              <ul>
                {sourceIndexCannotShrinkErrors.map((reason) => (
                  <li key={reason}>{reason}</li>
                ))}
              </ul>
              <EuiLink href={SHRINK_DOCUMENTATION_URL} target="_blank" rel="noopener noreferrer">
                Learn more
              </EuiLink>
            </div>
          </EuiCallOut>
          <EuiCallOut color="warning" hidden={sourceIndexCannotShrinkErrors.length != 0 || sourceIndexNotReadyReasons.length == 0}>
            <div style={{ lineHeight: 1.5 }}>
              <p>The source index is not ready to shrink, may due to the following reasons:</p>
              <ul>
                {sourceIndexNotReadyReasons.map((reason) => (
                  <li key={reason}>{reason}</li>
                ))}
              </ul>
              <EuiLink href={SHRINK_DOCUMENTATION_URL} target="_blank" rel="noopener noreferrer">
                Learn more
              </EuiLink>
            </div>
          </EuiCallOut>
          <EuiSpacer size="m" />
          <CustomFormRow label="Source index">
            <EuiFieldText value={this.props.sourceIndex.index} data-test-subj="sourceIndexForm" readOnly />
          </CustomFormRow>
          <EuiSpacer size="m" />
          <CustomFormRow
            label="Target index"
            helpText={"The name of the new shrunken index."}
            isInvalid={!!targetIndexNameError}
            error={targetIndexNameError}
          >
            <EuiFieldText
              value={targetIndexName}
              onChange={(e) => {
                this.onTargetIndexNameChange(e.target.value);
              }}
              isInvalid={!!targetIndexNameError}
              data-test-subj="targetIndexNameInput"
            />
          </CustomFormRow>
          <EuiSpacer size="m" />
          <FormGenerator
            ref={(ref) => (this.formRef = ref)}
            value={indexSettings}
            onChange={(value) => {
              this.setState({
                indexSettings: value,
              });
            }}
            formFields={formFields}
            hasAdvancedSettings
            advancedSettingsProps={{
              accordionProps: {
                initialIsOpen: false,
                id: "accordion_for_create_index_settings",
                buttonContent: <h4>Advanced settings</h4>,
              },
              rowProps: {
                label: "Specify advanced index settings",
                helpText: (
                  <>
                    Specify a comma-delimited list of settings.
                    <EuiLink href={INDEX_SETTINGS_URL} target="_blank">
                      View index settings
                    </EuiLink>
                  </>
                ),
              },
            }}
          />
        </EuiFlyoutBody>
        <EuiFlyoutFooter>
          <EuiFlexGroup justifyContent="flexEnd">
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty onClick={onClose} flush="left" data-test-subj="shrinkIndexCloseButton">
                Cancel
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton
                onClick={this.onClickAction}
                fill
                data-test-subj="shrinkIndexConfirmButton"
                disabled={sourceIndexCannotShrinkErrors.length != 0}
              >
                Shrink index
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlyoutFooter>
      </EuiFlyout>
    );
  }
}
