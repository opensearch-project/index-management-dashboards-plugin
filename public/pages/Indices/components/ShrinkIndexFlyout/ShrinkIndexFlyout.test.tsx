/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ShrinkIndexFlyout from "./ShrinkIndexFlyout";

describe("<ShrinkIndexFlyout /> spec", () => {
  it("renders the component", async () => {
    render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "0",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={() => {}}
        getIndexSettings={async () => {
          return {};
        }}
      />
    );

    await waitFor(() => {
      expect(document.body.children).toMatchSnapshot();
    });
  });

  it("calls close when cancel button clicked", async () => {
    const onClose = jest.fn();
    const { getByTestId } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "0",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={onClose}
        getIndexSettings={async () => {
          return {};
        }}
      />
    );

    await waitFor(() => {
      fireEvent.click(getByTestId("shrinkIndexCloseButton"));
      expect(onClose).toHaveBeenCalled();
    });
  });

  it("shows error when source index cannot shrink", async () => {
    const onClose = jest.fn();
    const setIndexSettings = jest.fn();
    const { getByTestId, queryByText } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "0",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={onClose}
        getIndexSettings={async () => {
          return {};
        }}
        setIndexSettings={setIndexSettings}
      />
    );

    await waitFor(async () => {
      expect(queryByText("The source index cannot shrink, due to the following reasons:")).not.toBeNull();
      expect(queryByText("The source index's write operations must be blocked.")).not.toBeNull();
      fireEvent.click(getByTestId("onSetIndexWriteBlockButton"));
      expect(setIndexSettings).toHaveBeenCalled();
    });
  });

  it("shows error when target index name is not set", async () => {
    const onClose = jest.fn();
    const testIndexSettings = {
      test_index: {
        settings: {
          "index.blocks.write": true,
          "index.routing.allocation.require._name": "node1",
        },
      },
    };
    const getIndexSettings = jest.fn().mockResolvedValue(testIndexSettings);
    const { getByTestId, queryByText } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "0",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={onClose}
        getIndexSettings={getIndexSettings}
      />
    );

    await waitFor(async () => {
      expect(queryByText("Target index name required.")).toBeNull();
    });

    await act(async () => {
      userEvent.type(getByTestId("targetIndexNameInput"), "test_index_shrunken");
    });
    await waitFor(async () => {
      expect(queryByText("Target index name required.")).toBeNull();
    });

    await act(async () => {
      userEvent.clear(getByTestId("targetIndexNameInput"));
    });
    await act(async () => {
      fireEvent.click(getByTestId("shrinkIndexConfirmButton"));
    });
    await waitFor(() => {
      expect(queryByText("Target index name required.")).not.toBeNull();
    });
  });

  it("shows error when number of replicas is not valid", async () => {
    const onClose = jest.fn();
    const testIndexSettings = {
      test_index: {
        settings: {
          "index.blocks.write": true,
          "index.routing.allocation.require._name": "node1",
        },
      },
    };
    const getIndexSettings = jest.fn().mockResolvedValue(testIndexSettings);
    const { getByTestId, queryByText } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "0",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={onClose}
        getIndexSettings={getIndexSettings}
      />
    );

    await waitFor(async () => {
      expect(queryByText("Number of replicas must be greater than or equal to 0.")).toBeNull();
    });

    await act(async () => {
      userEvent.clear(getByTestId("numberOfReplicasInput"));
    });

    await waitFor(async () => {
      expect(queryByText("Number of replicas must be greater than or equal to 0.")).not.toBeNull();
    });

    await act(async () => {
      userEvent.type(getByTestId("numberOfReplicasInput"), "-1");
    });

    await waitFor(async () => {
      expect(queryByText("Number of replicas must be greater than or equal to 0.")).not.toBeNull();
    });
  });

  it("shows danger when source index cannot shrink", async () => {
    const testIndexSettings = {
      test_index: {
        settings: {},
      },
    };
    const getIndexSettings = jest.fn().mockResolvedValue(testIndexSettings);
    const { getByTestId, queryByText } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "red",
          index: "test_index",
          pri: "1",
          rep: "1",
          status: "close",
        }}
        visible
        onConfirm={() => {}}
        onClose={() => {}}
        getIndexSettings={getIndexSettings}
      />
    );

    await waitFor(() => {
      expect(queryByText("The source index cannot shrink, due to the following reasons:")).not.toBeNull();
      expect(queryByText("The source index's health status is [red]!")).not.toBeNull();
      expect(queryByText("The source index has only one primary shard!")).not.toBeNull();
      expect(queryByText("The source index must be in open status!")).not.toBeNull();
      expect(getByTestId("shrinkIndexConfirmButton")).toHaveAttribute("disabled");
    });
  });

  it("shows warning when source index is not ready", async () => {
    const testIndexSettings = {
      test_index: {
        settings: {
          "index.blocks.read_only": true,
        },
      },
    };
    const getIndexSettings = jest.fn().mockResolvedValue(testIndexSettings);
    const { queryByText } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "yellow",
          index: "test_index",
          pri: "3",
          rep: "1",
          status: "open",
        }}
        visible
        onConfirm={() => {}}
        onClose={() => {}}
        getIndexSettings={getIndexSettings}
      />
    );

    await waitFor(() => {
      expect(queryByText("The source index is not ready to shrink, may due to the following reasons:")).not.toBeNull();
      expect(
        queryByText("Index setting [index.blocks.read_only] is [true], this will cause the new shrunken index's shards to be unassigned.")
      ).not.toBeNull();
      expect(queryByText("The source index's health is not green.")).not.toBeNull();
      expect(queryByText("One copy of every shard may not allocated to one node.")).not.toBeNull();
    });
  });

  it("no warning when source index is ready", async () => {
    const testIndexSettings = {
      test_index: {
        settings: {
          "index.blocks.write": true,
          "index.routing.allocation.require._name": "node1",
        },
      },
    };
    const getIndexSettings = jest.fn().mockResolvedValue(testIndexSettings);
    const onConfirm = jest.fn();
    const { queryByText, getByTestId } = render(
      <ShrinkIndexFlyout
        sourceIndex={{
          health: "green",
          index: "test_index",
          pri: "3",
          rep: "1",
          status: "open",
        }}
        visible
        onConfirm={onConfirm}
        onClose={() => {}}
        getIndexSettings={getIndexSettings}
      />
    );

    await waitFor(() => {
      expect(queryByText("The source index's health is not green.")).toBeNull();
      expect(queryByText("One copy of every shard may not allocated to one node.")).toBeNull();
    });

    await act(async () => {
      userEvent.type(getByTestId("targetIndexNameInput"), "test_index_shrunken");
    });
    await act(async () => {
      fireEvent.click(getByTestId("shrinkIndexConfirmButton"));
    });
    await waitFor(() => {
      expect(onConfirm).toHaveBeenCalled();
    });
  });
});
