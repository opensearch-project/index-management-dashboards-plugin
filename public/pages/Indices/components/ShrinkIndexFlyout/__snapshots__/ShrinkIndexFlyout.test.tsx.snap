// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShrinkIndexFlyout /> spec renders the component 1`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />,
  <div
    class="euiOverlayMask euiOverlayMask--belowHeader"
  >
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="euiFlyout euiFlyout--medium euiFlyout--paddingLarge"
        role="dialog"
        style="max-width: 600px;"
        tabindex="-1"
      >
        <div
          class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
        >
          <h2
            class="euiTitle euiTitle--medium"
            id="flyoutTitle"
          >
             Shrink index
          </h2>
        </div>
        <div
          class="euiFlyoutBody"
        >
          <div
            class="euiFlyoutBody__overflow"
            tabindex="0"
          >
            <div
              class="euiFlyoutBody__overflowContent"
            >
              <div
                class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
                style="padding-left: 0px; padding-right: 0px;"
              >
                <div
                  class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
                  style="padding: 0px 10px;"
                >
                  <div
                    class="euiFlexItem"
                  >
                    <h3
                      class="euiTitle euiTitle--small"
                    >
                      Source index details
                    </h3>
                  </div>
                </div>
                <hr
                  class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
                />
                <div
                  style="padding: 0px 10px;"
                >
                  <div
                    class="euiBasicTable euiBasicTable-loading"
                  >
                    <div>
                      <div
                        class="euiTableHeaderMobile"
                      >
                        <div
                          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                        >
                          <div
                            class="euiFlexItem euiFlexItem--flexGrowZero"
                          />
                          <div
                            class="euiFlexItem euiFlexItem--flexGrowZero"
                          />
                        </div>
                      </div>
                      <table
                        class="euiTable euiTable--responsive"
                        id="some_html_id"
                        tabindex="-1"
                      >
                        <caption
                          class="euiScreenReaderOnly euiTableCaption"
                        />
                        <thead>
                          <tr>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_index_0"
                              role="columnheader"
                              scope="col"
                            >
                              <span
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Index name"
                                >
                                  Index name
                                </span>
                              </span>
                            </th>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_pri_1"
                              role="columnheader"
                              scope="col"
                            >
                              <span
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Primary shards"
                                >
                                  Primary shards
                                </span>
                              </span>
                            </th>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_rep_2"
                              role="columnheader"
                              scope="col"
                            >
                              <span
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Replica shards"
                                >
                                  Replica shards
                                </span>
                              </span>
                            </th>
                            <th
                              class="euiTableHeaderCell"
                              data-test-subj="tableHeaderCell_store.size_3"
                              role="columnheader"
                              scope="col"
                            >
                              <span
                                class="euiTableCellContent"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                  title="Total index size"
                                >
                                  Total index size
                                </span>
                              </span>
                            </th>
                          </tr>
                        </thead>
                        <tbody>
                          <tr
                            class="euiTableRow"
                          >
                            <td
                              class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                              colspan="4"
                            >
                              <div
                                class="euiTableCellContent euiTableCellContent--alignCenter"
                              >
                                <span
                                  class="euiTableCellContent__text"
                                >
                                  No items found
                                </span>
                              </div>
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                  <div
                    class="euiSpacer euiSpacer--l"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlyoutFooter"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--flushLeft"
                data-test-subj="shrinkIndexCloseButton"
                type="button"
              >
                <span
                  class="euiButtonContent euiButtonEmpty__content"
                >
                  <span
                    class="euiButtonEmpty__text"
                  >
                    Cancel
                  </span>
                </span>
              </button>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButton euiButton--primary euiButton--fill euiButton-isDisabled"
                data-test-subj="shrinkIndexConfirmButton"
                disabled=""
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Shrink index
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;
